File: frontend/src/features/processing/presets.ts
import type {
  ProcessingControls,
  ProcessingManifest,
} from "../../lib/generated/contracts";
import type { ProcessingPresetOption } from "../../lib/types";

interface ProcessingUiState {
  defaultControls: ProcessingControls;
  presets: ProcessingPresetOption[];
}

const CUSTOM_PRESET_LABEL = "Custom";

const PRESET_COPY_BY_ID: Record<string, { label: string; description: string }> = {
  default: {
    label: "Neutral",
    description: "Balanced grayscale for a first review pass.",
  },
  xray: {
    label: "Bone Focus",
    description: "Bone palette with added punch and equalized detail.",
  },
  "high-contrast": {
    label: "High Contrast",
    description: "Sharper tonal separation without pseudocolor.",
  },
};

function toTitleCase(value: string): string {
  return value
    .split("-")
    .filter(Boolean)
    .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
    .join(" ");
}

function presetCopy(presetId: string): { label: string; description: string } {
  return (
    PRESET_COPY_BY_ID[presetId] ?? {
      label: toTitleCase(presetId),
      description: "Backend-defined processing preset.",
    }
  );
}

export function processingControlsEqual(
  left: ProcessingControls,
  right: ProcessingControls,
): boolean {
  // Sliders round contrast values in the UI, so allow a small tolerance before
  // deciding the current controls no longer match a preset exactly.
  return (
    left.brightness === right.brightness &&
    Math.abs(left.contrast - right.contrast) < 0.05 &&
    left.invert === right.invert &&
    left.equalize === right.equalize &&
    left.palette === right.palette
  );
}

export function buildProcessingUiState(
  manifest: ProcessingManifest,
): ProcessingUiState {
  const presets = manifest.presets.map((preset) => ({
    id: preset.id,
    controls: { ...preset.controls },
    ...presetCopy(preset.id),
  }));

  const defaultPreset =
    manifest.presets.find((preset) => preset.id === manifest.defaultPresetId) ??
    manifest.presets[0];

  return {
    defaultControls: defaultPreset
      ? { ...defaultPreset.controls }
      : {
          brightness: 0,
          contrast: 1.0,
          invert: false,
          equalize: false,
          palette: "none",
        },
    presets,
  };
}

export function matchPreset(
  controls: ProcessingControls,
  presets: readonly ProcessingPresetOption[],
): string {
  const matched = presets.find((preset) =>
    processingControlsEqual(preset.controls, controls),
  );

  return matched?.label ?? CUSTOM_PRESET_LABEL;
}

File: cmd/xrayview/presets.go
package main

import (
	"encoding/json"
	"io"
	"strings"
)

type processingControls struct {
	Brightness int     `json:"brightness"`
	Contrast   float64 `json:"contrast"`
	Invert     bool    `json:"invert"`
	Equalize   bool    `json:"equalize"`
	Palette    string  `json:"palette"`
}

type processingPreset struct {
	ID       string             `json:"id"`
	Controls processingControls `json:"controls"`
}

type processingManifest struct {
	DefaultPresetID string             `json:"defaultPresetId"`
	Presets         []processingPreset `json:"presets"`
}

const defaultPresetID = "default"

var processingPresets = []processingPreset{
	{
		ID: defaultPresetID,
		Controls: processingControls{
			Brightness: 0,
			Contrast:   1.0,
			Invert:     false,
			Equalize:   false,
			Palette:    "none",
		},
	},
	{
		ID: "xray",
		Controls: processingControls{
			Brightness: 10,
			Contrast:   1.4,
			Invert:     false,
			Equalize:   true,
			Palette:    "bone",
		},
	},
	{
		ID: "high-contrast",
		Controls: processingControls{
			Brightness: 0,
			Contrast:   1.8,
			Invert:     false,
			Equalize:   true,
			Palette:    "none",
		},
	},
}

func processingManifestData() processingManifest {
	presets := make([]processingPreset, len(processingPresets))
	copy(presets, processingPresets)

	return processingManifest{
		DefaultPresetID: defaultPresetID,
		Presets:         presets,
	}
}

func writeProcessingManifest(w io.Writer) error {
	encoder := json.NewEncoder(w)
	return encoder.Encode(processingManifestData())
}

func lookupProcessingPreset(id string) (processingPreset, bool) {
	for _, preset := range processingPresets {
		if preset.ID == id {
			return preset, true
		}
	}

	return processingPreset{}, false
}

func supportedPresetIDs() []string {
	ids := make([]string, 0, len(processingPresets))
	for _, preset := range processingPresets {
		ids = append(ids, preset.ID)
	}

	return ids
}

func supportedPresetList() string {
	return strings.Join(supportedPresetIDs(), ", ")
}
